### PR TITLE
Fix exception in segmented-reduce benchmark

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes exception thrown in `REDUCTION_NVBENCH`. 

```
Run:  [121/768] segmented_reduction_simple [Device=0 InputType=bool OutputType=F32 AggregationKinds=2 column_size=100000 num_segments=1000]
Fail: Unexpected error: cuDF failure at: /cudf/cpp/src/reductions/segmented_min.cu:33: segmented_min() operation requires matching output type
```
(There are 144 of these)

The code is refactored to reduce the generated benchmark code in order to match input and output types and to force the output type to `BOOL8` for aggregation kind `ALL`. This also removes code generated that is only _skipped_:

```
Run:  [277/768] segmented_reduction_simple [Device=0 InputType=I32 OutputType=I32 AggregationKinds=7 column_size=100000 num_segments=1000]
Skip: Invalid combination of dtype and aggregation type.

```
(There are also 144 of these)

So 37.5% (288/768) of the benchmark runs were either invalid or throw an exception

The code change reduces the benchmarks to 192 valid ones.

Matching type input/output types should provide sufficient coverage since internally the numeric reductions `SUM` and `PRODUCT` are performed on `int64` or `double` and then the results are cast to the output-type. So the `int32` and `float` types will always cover this cast step. Extra measurements for casting to `double` to `int32` for example should not be necessary.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/reduction/segment_reduce.cu [hunk 1]
```diff
@@ -32,99 +32,98 @@
...
+  auto const offset_span = cudf::device_span<cudf::size_type>{offsets};
 
   state.exec(
-    nvbench::exec_tag::sync,
-    [input_view = input->view(), offset_span = device_span<size_type>{offsets}, &agg](
-      nvbench::launch& launch) {
-      segmented_reduce(
-        input_view, offset_span, *agg, data_type{type_to_id<OutputType>()}, null_policy::INCLUDE);
+    nvbench::exec_tag::sync, [input_view, output_type, offset_span, &agg](nvbench::launch& launch) {
+      segmented_reduce(input_view, offset_span, *agg, output_type, cudf::null_policy::INCLUDE);
```
Comment (PointKernel) [inline cpp/benchmarks/reduction/segment_reduce.cu:114]: Similar to the implementation here: https://github.com/rapidsai/cudf/blob/63a47d91fc558a5b0b89f3ce9b0a61ffb3ed84f5/cpp/benchmarks/groupby/group_max.cpp#L55-L57
We need to pass the target stream to nvbench `state` thus performance metrics can be recorded properly. I'm a bit surprised that the execution didn't crash with the current implementation.

**Review by PointKernel — CHANGES_REQUESTED** [review]
768 -> 192 :fire: 

Only one small issue when benchmarking with implicit stream.

**Review by bdice — APPROVED** [review]
Thanks @davidwendt! Please add the stream as noted by @PointKernel, otherwise LGTM.

**Review by PointKernel — APPROVED** [review]
LGTM

**Review by ttnghia — APPROVED** [review]

**Pull request merged**